FBSnapshotTestCase/Categories/UIImage+Compare.m

```diff
@@ -129,4 +115,63 @@ - (BOOL)fb_compareWithImage:(UIImage *)image tolerance:(CGFloat)tolerance
     return imageEqual;
 }
 
+- (BOOL)_comparePixelWithPixelTolerance:(CGFloat)pixelTolerance
+                         referencePixel:(FBComparePixel *)referencePixel
+                             imagePixel:(FBComparePixel *)imagePixel
+{
+    if (referencePixel->raw == imagePixel->raw) {
+        return YES;
+    } else if (pixelTolerance == 0) {
+        return NO;
+    }
+
+    CGFloat redPercentDiff = [self _calculatePercentDifferenceForReferencePixelComponent:referencePixel->pixels.red
+                                                                     imagePixelComponent:imagePixel->pixels.red];
+    CGFloat greenPercentDiff = [self _calculatePercentDifferenceForReferencePixelComponent:referencePixel->pixels.green
+                                                                       imagePixelComponent:imagePixel->pixels.green];
+    CGFloat bluePercentDiff = [self _calculatePercentDifferenceForReferencePixelComponent:referencePixel->pixels.blue
+                                                                      imagePixelComponent:imagePixel->pixels.blue];
+    CGFloat alphaPercentDiff = [self _calculatePercentDifferenceForReferencePixelComponent:referencePixel->pixels.alpha
+                                                                       imagePixelComponent:imagePixel->pixels.alpha];
+
+    return !(redPercentDiff > pixelTolerance ||
+             greenPercentDiff > pixelTolerance ||
+             bluePercentDiff > pixelTolerance ||
+             alphaPercentDiff > pixelTolerance);
+}
+
+- (CGFloat)_calculatePercentDifferenceForReferencePixelComponent:(char)p1
+                                             imagePixelComponent:(char)p2
+{
+    int referencePixelComponent = (unsigned char)p1;
+    int imagePixelComponent = (unsigned char)p2;
+    int componentDifference = abs(referencePixelComponent - imagePixelComponent);
+    return (CGFloat)componentDifference / 256;
+}
+
+- (BOOL)_comparePixelWithPixelTolerance:(CGFloat)pixelTolerance
+                              tolerance:(CGFloat)tolerance
+                             pixelCount:(NSInteger)pixelCount
+                        referencePixels:(FBComparePixel *)referencePixel
+                            imagePixels:(FBComparePixel *)imagePixel
+{
+    NSInteger numDiffPixels = 0;
+    for (NSInteger n = 0; n < pixelCount; ++n) {
+        // If this pixel is different, increment the pixel diff count and see
+        // if we have hit our limit.
+        if (![self _comparePixelWithPixelTolerance:pixelTolerance referencePixel:referencePixel imagePixel:imagePixel]) {
+            numDiffPixels++;
+
+            CGFloat percent = (CGFloat)numDiffPixels / pixelCount;
```

Comment:
just to be clearer; `CGFloat percent = (CGFloat)(numDiffPixels / pixelCount);`

Reply:
Done 👍

Reply:
We want to make the cast before the calculation happens so the result is a `CGFloat`, otherwise small percentages will be rounded to zero if the calculation happens with `NSInteger`.

I'll change It to `CGFloat percent = (CGFloat)numDiffPixels / (CGFloat)pixelCount;` to make it clearer